diff -up libvirt-0.7.6/tests/Makefile.am.jx libvirt-0.7.6/tests/Makefile.am
--- libvirt-0.7.6/tests/Makefile.am.jx	2010-01-21 09:32:37.000000000 -0500
+++ libvirt-0.7.6/tests/Makefile.am	2010-02-17 10:32:27.000000000 -0500
@@ -37,7 +37,8 @@ LDADDS = \
 	../src/libvirt_test.la \
 	../gnulib/lib/libgnu.la \
 	$(LIBSOCKET) \
-        $(COVERAGE_LDFLAGS)
+        $(COVERAGE_LDFLAGS) \
+	-ldl
 
 qemuhelpdata = \
 	kvm-74 \
